### PR TITLE
chore: dispatch all events to EventBus

### PR DESCRIPTION
**Pull request by williazz**
> Rewrite of #440 

### Purpose

This PR creates the structure necessary for sharing information between plugins. 

### Uses cases: 

1. Have PluginA find eventId's created by PluginB etc. Useful for LCP attribution
2. WebVitalsPlugin tracks current count of in-flight Xhr and Fetch requests

### Revisions

**Revision 1**

I have created `EventBus` with topic-based filtering. I implemented our own instead of taking on a dependency because we only need a few extremely simple features. 
* Plugins can subscribe to any topic via `context.bus.subscribe(topic, callback)`
* Plugins can publish to any topic via `context.bus.dispatch(topic, message)`
* All events are published on every record by EventCache via `bus.dispatch(eventType, message)`

**Revision 2**
* Plugins are responsible for dispatching events
* EventCache is not responsible for dispatching events
* EventCache.record returns a ParsedRumEvent so plugins can access things like eventId, metadata, details, etc. 
* Read this thread for full context https://github.com/aws-observability/aws-rum-web/pull/445#discussion_r1317539603

**Revision 3**
* EventCache dispatches every event that is successfully recorded to topic 'events'
* Resource event supplies additional key=PerformanceResourceTiming
* Navigation supplies additional key=PerformanceNavigationTiming for level 2
* Read for full context https://github.com/aws-observability/aws-rum-web/pull/445#discussion_r1317539603

**Revision 4**
* Remove key
* Target main branch
### Examples 

**Use case 1:** PluginA finds eventId's of events from PluginB
```typescript
class WebVitalsPlugin {
   imageEvents = []
   constructor() {
      this.context.bus.subscribe(PERFORMANCE_RESOURCE_EVENT, this.storeResourceEvent)
   }
   storeResourceEvent(event) {
      if (event.fileType === ResourceTypes.IMAGE) {
         this.imageEvents.push(event)
      }
   }

   ...
   onLCP(payload => {
       ...
       findMatchingLCPEvent()
       // teardown
       this.imageEvents = []
       this.context.bus.unsubscribe(...)
   })
}
```

**Use case 2:** Count in-flight fetch/xhr requests
```typescript
// FetchPlugin.ts
...
private fetch() {
   ...
   this.context.bus.dispatch('http_open_count', {payload: 1})
   return original()
   .apply()
   .then(() => {
        this.context.bus.dispatch('http_open_count', {payload: -1})
    })
   .catch(() => {
        this.context.bus.dispatch('http_open_count', {payload: -1})
    })
}

...

// WebVitalsPlugin
class WebVitalsPlugins {
    ...
    http_open_count = 0
    handleHttpCountChange({payload}) { this.http_open_count += payload }
    constructor() {
        this.context.bus.subscribe('http_open_count', this.handleHttpCountChange)
   }
   teardown() {
       this.context.bus.unsubscribe('http_open_count', this.handleHttpCountChange)
   }
   ....
}
```

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/event-cache/EventCache.ts [hunk 1]
```diff
@@ -220,12 +225,20 @@ export class EventCache {
...
 
-        this.events.push({
-            details: JSON.stringify(eventData),
+        const partialEvent = {
             id: v4(),
-            metadata: JSON.stringify(metaData),
             timestamp: new Date(),
             type
+        };
+        this.bus.notify(type, {
```
Comment (williazz) [inline src/event-cache/EventCache.ts:233]: Note: The "Raw" RumEvent is published so that plugins do not have to de-stringify the information they want to look at.

File: src/orchestration/Orchestration.ts [hunk 2]
```diff
@@ -444,7 +446,8 @@ export class Orchestration {
             config: this.config,
             record: this.eventCache.recordEvent,
             recordPageView: this.eventCache.recordPageView,
-            getSession: this.eventCache.getSession
+            getSession: this.eventCache.getSession,
+            bus: this.bus
```
Comment (williazz) [inline src/orchestration/Orchestration.ts:450]: Nit: Also open to flattening the bus 
```
context = {
...
subscribe: ...,
unsubscribe: ...,
notify: ...,
}
```

File: src/plugins/event-plugins/__tests__/FetchPlugin.test.ts [hunk 3]
```diff
@@ -542,17 +545,9 @@ describe('FetchPlugin tests', () => {
...
-        const xRayOnContext: PluginContext = {
-            applicationId: 'b',
-            applicationVersion: '1.0',
-            config: { ...DEFAULT_CONFIG, ...{ enableXRay: true } },
-            record,
-            recordPageView,
-            getSession
-        };
-
+        const context = Object.assign({}, xRayOnContext, { getSession });
```
Comment (williazz) [inline src/plugins/event-plugins/__tests__/FetchPlugin.test.ts:548]: fix unrelated breaking changes

File: src/plugins/event-plugins/__tests__/ResourcePlugin.test.ts [hunk 4]
```diff
@@ -70,18 +70,14 @@ describe('ResourcePlugin tests', () => {
...
         mockRandom(0); // Retain order in shuffle
-        const context: PluginContext = {
-            applicationId: 'b',
-            applicationVersion: '1.0',
-            config: { ...DEFAULT_CONFIG, recordResourceUrl: false },
-            record,
-            recordPageView,
-            getSession
-        };
+
```
Comment (williazz) [inline src/plugins/event-plugins/__tests__/ResourcePlugin.test.ts:73]: fix unrelated breaking changes

File: src/plugins/event-plugins/__tests__/XhrPlugin.test.ts [hunk 5]
```diff
@@ -583,14 +579,7 @@ describe('XhrPlugin tests', () => {
...
         const getSession: jest.MockedFunction<GetSession> = jest.fn();
-        const context: PluginContext = {
-            applicationId: 'b',
-            applicationVersion: '1.0',
-            config: { ...DEFAULT_CONFIG, ...{ enableXRay: true } },
-            record,
-            recordPageView,
-            getSession
-        };
+        const context = { ...mockContext, getSession };
```
Comment (williazz) [inline src/plugins/event-plugins/__tests__/XhrPlugin.test.ts:582]: fix unrelated breaking changes

File: src/plugins/types.ts [hunk 6]
```diff
@@ -13,4 +13,5 @@ export type PluginContext = {
     record: RecordEvent;
     recordPageView: RecordPageView;
     getSession: GetSession;
+    bus: EventBus;
```
Comment (williazz) [inline src/plugins/types.ts:16]: Nit: Also open to flattening the bus 
```
context = {
...
subscribe: ...,
unsubscribe: ...,
notify: ...,
}
```

File: src/event-cache/EventCache.ts (hunk 1 above)
Comment (qhanam) [reply src/event-cache/EventCache.ts:233]: question: We are potentially publishing more than needed. Consider the following options.
1. The EventCache publishes whenever any event is recorded.
2. The Resource plugin publishes whenever it records a PerformanceResource event.

Comment (williazz) [reply src/event-cache/EventCache.ts:233]: I can refactor so that record() returns a raw rum event, which includes fields like `id` and `metadata`. That way, ResourcePlugin can publish all the information we need, for instance. 

Do you have any issue with the publishing parsed rum event?

Comment (qhanam) [reply src/event-cache/EventCache.ts:233]: Does the web vitals plugin need a reference to the PerformanceResource object? If so, I'm not sure (1) will work.

Comment (williazz) [reply src/event-cache/EventCache.ts]: (1) is a simpler implementation

> Does the web vitals plugin need a reference to the PerformanceResource object?

It does not need PerformanceResourceTiming object as a key because the URL is also a good key. But the performance object is the best key

File: src/plugins/event-plugins/DemoPlugin.ts [hunk 7]
```diff
@@ -19,7 +19,7 @@ export class DemoPlugin implements Plugin {
         this.configuration = {};
         this.timerId = undefined;
         // eslint-disable-next-line @typescript-eslint/no-empty-function
-        this.recordEvent = () => {};
+        this.recordEvent = () => undefined;
```
Comment (williazz) [inline src/plugins/event-plugins/DemoPlugin.ts:22]: note: linter was complaining about this one

File: src/plugins/event-plugins/NavigationPlugin.ts [hunk 8]
```diff
@@ -292,4 +282,19 @@ export class NavigationPlugin extends InternalPlugin {
...
     }
+
+    private dispatchedLoad = false;
+    /** Record event and publish if record was successful */
+    private recordEvent(event: NavigationEvent) {
+        const parsedEvent = this.context?.record(
+            PERFORMANCE_NAVIGATION_EVENT_TYPE,
+            event
+        );
+        if (parsedEvent && !this.dispatchedLoad) {
```
Comment (williazz) [inline src/plugins/event-plugins/NavigationPlugin.ts:293]: Note: Only first event is needed

File: src/event-bus/__tests__/EventBus.test.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+describe('EventBus tests', () => {
+    let eventBus: EventBus;
+    const l1 = jest.fn();
+    const l2 = jest.fn();
+    beforeEach(() => {
+        eventBus = new EventBus();
+        jest.clearAllMocks();
+    });
+    test('when dispatch is invoked then all listeners are called', async () => {
```
Comment (ps863) [inline src/event-bus/__tests__/EventBus.test.ts:11]: nit: consider adding test to subscribe to different topics

File: src/event-bus/EventBus.ts [hunk 10]
```diff
@@ -0,0 +1,41 @@
+export type Subscriber = (message: any) => void;
+export interface Message {
+    key?: any;
+    payload: any;
+}
+
+/** A topic-based event bus to facilitate communication between plugins */
+export default class EventBus {
```
[the remark this answers is not in the document]
Comment (williazz) [reply src/event-bus/EventBus.ts:8]: Sorry, I renamed it to dispatch since notify was a bad name.

Notify is usually to notify of state changes that have already happened. Dispatch is usually for sending messages. So I thought dispatch was a better name

File: src/event-cache/EventCache.ts [hunk 11]
```diff
@@ -82,8 +88,12 @@ export class EventCache {
      * If the session is not being recorded, the event will not be recorded.
      *
      * @param type The event schema.
+     * @returns {ParsedRumEvent | undefined} if the event was recorded
      */
-    public recordEvent = (type: string, eventData: object) => {
+    public recordEvent = (
+        type: string,
+        eventData: object
+    ): ParsedRumEvent | undefined => {
```
Comment (williazz) [inline src/event-cache/EventCache.ts:96]: see https://github.com/aws-observability/aws-rum-web/pull/445#discussion_r1316495440

**Comment by ps863** [comment]
high level question: could the addition and usage of an event bus impact the memory/performance footprint of the web client on web applications?

**Comment by williazz** [comment]
> high level question: could the addition and usage of an event bus impact the memory/performance footprint of the web client on web applications?

We are currently only monitoring the size and relying design principles to limit memory and cpu usage. I have also been thinking about how to objectively monitor this performance. 

But from a design perspective, I think the impact is minimal.

File: src/event-cache/EventCache.ts (hunk 1 above)
Comment (ps863) [reply src/event-cache/EventCache.ts]: +1. I think (1) makes more sense for the points made by @williazz

Comment (qhanam) [reply src/event-cache/EventCache.ts]: > +1. I think (1) makes more sense for the points made by @williazz

I agree.

> If we go back to (1), we can just add an optional key parameter to record() and have EventCache dispatch both key and payload.

Could we set the topic to 'event' and let the listener decide which events to do something with?

**Comment by qhanam** [comment]
My thoughts on (1) vs (2):
**If doing (1)**, this topic allows any plugin to listen for any event. Publish all events under an ‘events’ topic.
**If doing (2)**, this topic is for the specific purpose of tracking resource/nav event Ids.  Return the event id from recordEvent and publish an <event id, url> tuple.

File: src/event-bus/__tests__/EventBus.test.ts (hunk 9 above)
Comment (williazz) [reply src/event-bus/__tests__/EventBus.test.ts:15]: Added an extra test

File: src/event-cache/EventCache.ts [hunk 12]
```diff
@@ -220,12 +225,23 @@ export class EventCache {
...
-        this.events.push({
-            details: JSON.stringify(eventData),
+        const partialEvent = {
             id: v4(),
-            metadata: JSON.stringify(metaData),
             timestamp: new Date(),
             type
+        };
+        this.eventBus.dispatch(Topic.EVENTS, {
+            ...(key && { key }),
```
Comment (williazz) [inline src/event-cache/EventCache.ts:234]: Note: `...(key && { key })` checks that key exists before adding property, avoiding dispatching the unnecessary `{key: undefined}`

File: src/event-cache/EventCache.ts (hunk 1 above)
Comment (williazz) [reply src/event-cache/EventCache.ts]: > Could we set the topic to 'event' and let the listener decide which events to do something with?

Done in revision 3

**Review by qhanam — CHANGES_REQUESTED** [review]

File: src/event-bus/EventBus.ts [hunk 13]
```diff
@@ -0,0 +1,44 @@
+export type Subscriber = (message: Message) => void;
+export interface Message {
+    key?: any;
```
[the remark this answers is not in the document]
Comment (williazz) [reply src/event-bus/EventBus.ts]: I have removed key

File: src/plugins/types.ts (hunk 6 above)
Comment (williazz) [reply src/plugins/types.ts]: Renamed to eventBus

File: src/plugins/event-plugins/NavigationPlugin.ts [hunk 14]
```diff
@@ -265,7 +265,8 @@ export class NavigationPlugin extends InternalPlugin {
         if (this.context?.record) {
             this.context.record(
                 PERFORMANCE_NAVIGATION_EVENT_TYPE,
-                eventDataNavigationTimingLevel2
+                eventDataNavigationTimingLevel2,
+                entryData
```
[the remark this answers is not in the document]
Comment (williazz) [reply src/plugins/event-plugins/NavigationPlugin.ts]: I have removed these changes

File: src/event-bus/EventBus.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (williazz) [reply src/event-bus/EventBus.ts]: I agree that `key` does not serve a general purpose and have removed it. 

One drawback of (1) or (2) is that it requires plugins to execute a linear search on `performance.getEntries()` to find a matching hash or tuple. Or to maintain a map. With `key`, they get the matching entry directly.

File: src/event-cache/EventCache.ts [hunk 15]
```diff
@@ -220,12 +225,22 @@ export class EventCache {
...
             timestamp: new Date(),
             type
+        };
+        this.eventBus.dispatch(Topic.EVENT, {
+            payload: {
+                ...partialEvent,
+                details: eventData,
+                metadata: metaData
+            }
+        });
```
Comment (qhanam) [inline src/event-cache/EventCache.ts:239]:
```suggestion
        this.eventBus.dispatch(Topic.EVENT, {
            ...partialEvent,
            details: eventData,
            metadata: metaData
        });
```

File: src/plugins/event-plugins/__tests__/NavigationPlugin.test.ts [hunk 16]
```diff
@@ -77,6 +77,9 @@ describe('NavigationPlugin tests', () => {
                 navigationTimingLevel: 1
             })
         );
+
+        // // expect to record without key
+        expect(record.mock.calls[0].length).toEqual(2);
```
Comment (qhanam) [inline src/plugins/event-plugins/__tests__/NavigationPlugin.test.ts:82]:
```suggestion
```

File: src/plugins/event-plugins/__tests__/FetchPlugin.test.ts (hunk 3 above)
Comment (qhanam) [reply src/plugins/event-plugins/__tests__/FetchPlugin.test.ts:543]: What was the problem here?

File: src/event-cache/EventCache.ts [hunk 17]
```diff
@@ -198,7 +203,7 @@ export class EventCache {
      *
      * @param type The event schema.
      */
-    private addRecordToCache = (type: string, eventData: object) => {
+    private addRecordToCache = (type: string, eventData: object, key?: any) => {
```
Comment (qhanam) [inline src/event-cache/EventCache.ts:206]:
```suggestion
    private addRecordToCache = (type: string, eventData: object) => {
```

**Review by qhanam — APPROVED** [review]

File: src/event-bus/EventBus.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (williazz) [reply src/event-bus/EventBus.ts]: Issue: 

(1) we cannot reliably create a `id` from <name, entryType, startTime, duration> because `recordResourceURL` is configurable. If turned off, we cannot match `PerformanceEntry.name` to `ResourceEvent.targetUrl` because the latter does not exist. If we do (1), it would have to use id=<entryType, startTime, duration>, which is still a decent key. 

(2) is the same thing as dispatching `{key: performanceEntry, payload:...}`, since the object hash is implied in javascript. Manually creating the hash would require [object-hash](https://github.com/puleos/object-hash) (~16M weekly downloads) but has size=~60kb. Doing our own implementation will be smaller but not very performant. 

In my opinion, if the goal is for key(performanceEntry) === key(resourceEvent), then **the most serviceable option is option (1) with key=<entryType, startTime, duration>**
